### PR TITLE
add suggestion bundle support

### PR DESCRIPTION
**Pull request by granthamtaylor**
## Why are the changes needed?

I recently created the `flytekitplugins.optuna` plugin.

This PR adds support to bundle all of the suggestions into any number of dictionaries.

This is for the cases in which one wants to add dozens or more hyperparameters. Adding each of them as a type function input is pretty exhausting. In our case, we can collect all of these suggestions into any number of dictionaries at runtime.

## What changes were proposed in this pull request?

I add support for bundled suggestions to `flytekitplugins.optuna`. Users should provide any kwargs of type `dict[str, Suggestion|Any]`.

## How was this patch tested?

I have added an example and unit tests for the new bundling functionality.

### Setup process

```python
import math
from typing import Union

import flytekit as fl
from flytekitplugins.optuna import Optimizer, suggest

@fl.task(container_image=image)
async def objective(suggestions: dict[str, Union[int, float]], power: int) -> float:

    # building out a large set of typed inputs is exhausting, so we can just use a dict

    x, y, z = suggestions["x"], suggestions["y"], suggestions["z"]

    return math.log((((x - 5) ** 2) + (y + 4) ** 4 + (3 * z - 3) ** 2)) ** power


@fl.eager(container_image=image)
async def train(concurrency: int, n_trials: int) -> float:
    optimizer = Optimizer(objective, concurrency, n_trials)

    suggestions = {
        "x": suggest.float(low=-10, high=10),
        "y": suggest.integer(low=-10, high=10),
        "z": suggest.category([-5, 0, 3, 6, 9]),
    }

    await optimizer(suggestions=suggestions, power=2)

    return optimizer.study.best_value
```

## Check all the applicable boxes 

- [x] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off. 
 <div id='description'>
<h3>Summary by Bito</h3>
Enhanced flytekit-optuna plugin with bundled hyperparameter suggestions and improved type safety. Implemented dictionary-based parameter passing and callback mechanism for better management. Modified Optimizer class with instance-based suggest method and parameter organization using parent-child relationships. Added proper handling of optimizer's best value and streamlined multiple hyperparameter handling through runtime dictionary collection.
<br>
<br>
<b>Unit tests added</b>: True
<br>
<br>
<b>Estimated effort to review (1-5, lower is better)</b>: 2
</div>

### CODE REVIEW

**Comment by granthamtaylor** [comment]
Another example:


```python
import asyncio

import flytekit as fl
import sklearn
from sklearn.model_selection import train_test_split
import xgboost as xgb
import numpy as np
import optuna
import plotly

from flytekitplugins.optuna import Optimizer, suggest

image = fl.ImageSpec(
    builder="union",
    packages=[
        "flytekit==1.15.0b0",
        "optuna>=4.0.0",
        "scikit-learn>=1.6.0",
        "xgboost>=2.1.3",
        "plotly>=5.24.1",
    ],
)


@fl.task(container_image=image)
async def objective(params: dict[str, int | float | str]) -> float:

    (data, target) = sklearn.datasets.load_breast_cancer(return_X_y=True)
    train_x, valid_x, train_y, valid_y = train_test_split(data, target, test_size=0.25)
    dtrain = xgb.DMatrix(train_x, label=train_y)
    dvalid = xgb.DMatrix(valid_x, label=valid_y)

    bst = xgb.train(params, dtrain)
    valid_yhat = bst.predict(dvalid)
    accuracy = sklearn.metrics.accuracy_score(valid_y, np.rint(valid_yhat))

    return accuracy


@fl.eager(container_image=image)
async def train(concurrency: int, n_trials: int):
    
    study = optuna.create_study(direction="maximize")
    
    optimizer = Optimizer(objective, concurrency, n_trials, study)

    params = {
        "lambda": suggest.float(1e-8, 1.0, log=True),
        "alpha": suggest.float(1e-8, 1.0, log=True),
        "subsample": suggest.float(0.2, 1.0),
        "colsample_bytree": suggest.float(0.2, 1.0),
        "max_depth": suggest.integer(3, 9, step=2),
        "min_child_weight": suggest.integer(2, 10),
        "eta": suggest.float(1e-8, 1.0, log=True),
        "gamma": suggest.float(1e-8, 1.0, log=True),
        "grow_policy": suggest.category(["depthwise", "lossguide"]),
        "sample_type": suggest.category(["uniform", "weighted"]),
        "normalize_type": suggest.category(["tree", "forest"]),
        "rate_drop": suggest.float(1e-8, 1.0, log=True),
        "skip_drop": suggest.float(1e-8, 1.0, log=True),
        "objective": "binary:logistic",
        "tree_method": "exact",
        "booster": "dart",
    }

    await optimizer(params=params)
```

**Comment by granthamtaylor** [comment]
In some extreme cases, it is useful to be able to define suggestions programmatically. However, this cannot happen within an objective task, because we need to tell the Study what each task is doing. Therefore, users should be able to define a `callback` function in which they may express their programmatic suggestion logic.

This is _complicated_ but I think it is the best solution. It takes advantage of Flyte's static typing and enables caching too.

```python
import asyncio
from typing import Any

import flytekit as fl
import sklearn
from sklearn.model_selection import train_test_split
import xgboost as xgb
import numpy as np
import optuna
import plotly

from optimizer import Optimizer

image = fl.ImageSpec(
    builder="union",
    packages=[
        "flytekit==1.15.0b0",
        "optuna>=4.0.0",
        "scikit-learn>=1.6.0",
        "xgboost>=2.1.3",
        "plotly>=5.24.1",
    ],
)


@fl.task(container_image=image)
async def objective(params: dict[str, int | float | str]) -> float:
    (data, target) = sklearn.datasets.load_breast_cancer(return_X_y=True)
    train_x, valid_x, train_y, valid_y = train_test_split(data, target, test_size=0.25)
    dtrain = xgb.DMatrix(train_x, label=train_y)
    dvalid = xgb.DMatrix(valid_x, label=valid_y)

    bst = xgb.train(params, dtrain)
    valid_yhat = bst.predict(dvalid)
    accuracy = sklearn.metrics.accuracy_score(valid_y, np.rint(valid_yhat))
    
    print(accuracy)

    return accuracy


@fl.eager(container_image=image)
async def train(concurrency: int, n_trials: int):

    study = optuna.create_study(direction="maximize")
    
    def callback(trial: optuna.Trial, inputs: dict[str, Any]) -> dict[str, Any]:

        params = {
            "verbosity": 0,
            "objective": "binary:logistic",
            # use exact for small dataset.
            "tree_method": "exact",
            # defines booster, gblinear for linear functions.
            "booster": trial.suggest_categorical("booster", ["gbtree", "gblinear", "dart"]),
            # L2 regularization weight.
            "lambda": trial.suggest_float("lambda", 1e-8, 1.0, log=True),
            # L1 regularization weight.
            "alpha": trial.suggest_float("alpha", 1e-8, 1.0, log=True),
            # sampling ratio for training data.
            "subsample": trial.suggest_float("subsample", 0.2, 1.0),
            # sampling according to each tree.
            "colsample_bytree": trial.suggest_float("colsample_bytree", 0.2, 1.0),
        }

        if params["booster"] in ["gbtree", "dart"]:
            # maximum depth of the tree, signifies complexity of the tree.
            params["max_depth"] = trial.suggest_int("max_depth", 3, 9, step=2)
            # minimum child weight, larger the term more conservative the tree.
            params["min_child_weight"] = trial.suggest_int("min_child_weight", 2, 10)
            params["eta"] = trial.suggest_float("eta", 1e-8, 1.0, log=True)
            # defines how selective algorithm is.
            params["gamma"] = trial.suggest_float("gamma", 1e-8, 1.0, log=True)
            params["grow_policy"] = trial.suggest_categorical("grow_policy", ["depthwise", "lossguide"])

        if params["booster"] == "dart":
            params["sample_type"] = trial.suggest_categorical("sample_type", ["uniform", "weighted"])
            params["normalize_type"] = trial.suggest_categorical("normalize_type", ["tree", "forest"])
            params["rate_drop"] = trial.suggest_float("rate_drop", 1e-8, 1.0, log=True)
            params["skip_drop"] = trial.suggest_float("skip_drop", 1e-8, 1.0, log=True)

        return dict(params=params)

    optimizer = Optimizer(objective, concurrency, n_trials, study, callback)

    await optimizer()

    show("History", optuna.visualization.plot_optimization_history(optimizer.study))
    show("Timeline", optuna.visualization.plot_timeline(optimizer.study))
    show("Coordinates", optuna.visualization.plot_parallel_coordinate(optimizer.study))
    show("Importance", optuna.visualization.plot_param_importances(optimizer.study))


def show(name: str, fig: plotly.graph_objects.Figure):
    html: str = plotly.io.to_html(
        fig=fig,
        config={
            "scrollZoom": False,
            "displayModeBar": False,
            "doubleClick": "reset",
        },
    )

    fl.Deck(name, html)


if __name__ == "__main__":
    loss = asyncio.run(train(concurrency=1, n_trials=10))
    print(loss)

```
